sql/core/src/main/scala/org/apache/spark/sql/execution/systemcatalog/InformationSchema.scala

```diff
@@ -0,0 +1,230 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.systemcatalog
+
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql._
+import org.apache.spark.sql.catalyst.catalog.CatalogDatabase
+import org.apache.spark.sql.catalyst.dsl.plans._
+import org.apache.spark.sql.catalyst.expressions.Alias
+import org.apache.spark.sql.catalyst.plans.logical.Project
+import org.apache.spark.sql.execution.datasources._
+import org.apache.spark.sql.sources._
+import org.apache.spark.sql.types._
+
+object InformationSchema extends Logging {
+  def register(sparkSession: SparkSession, relation: BaseRelation, names: Seq[String]) {
+    val plan = LogicalRelation(relation).analyze
+    val projectList = plan.output.zip(plan.schema).map {
+      case (attr, col) => Alias(attr, col.name)()
+    }
+    sparkSession.sessionState.executePlan(Project(projectList, plan))
+    for (name <- names)
+      sparkSession.sessionState.catalog.createTempView(name, plan, overrideIfExists = true)
+  }
+
+  def build(sparkSession: SparkSession) {
+    // Create database
+    val informationSchemaDb = new CatalogDatabase(
+      "information_schema",
+      "Spark information schema",
+      locationUri = "/tmp/",
+      Map())
+    sparkSession.sessionState.catalog.createDatabase(informationSchemaDb, ignoreIfExists = true)
+
+    // Create tables
+    register(sparkSession, new DatabasesRelation(sparkSession), Seq("schemata", "databases"))
+    register(sparkSession, new TablesRelation(sparkSession), Seq("tables"))
+    register(sparkSession, new ViewsRelation(sparkSession), Seq("views"))
+    register(sparkSession, new ColumnsRelation(sparkSession), Seq("columns"))
+    register(sparkSession, new SessionVariablesRelation(sparkSession), Seq("session_variables"))
+  }
+}
+
+class DatabasesSource extends SchemaRelationProvider {
+  override def createRelation(
+      sqlContext: SQLContext,
+      parameters: Map[String, String],
+      schema: StructType): BaseRelation = {
+    DatabasesRelation(sqlContext.sparkSession)
+  }
+}
+
+case class DatabasesRelation(@transient sparkSession: SparkSession)
+  extends BaseRelation with PrunedFilteredScan {
+
+  override def sqlContext: SQLContext = sparkSession.sqlContext
+
+  override def schema: StructType = StructType(Seq(
+    StructField("CATALOG_NAME", StringType, false),
+    StructField("SCHEMA_NAME", StringType, false)
+  ))
+
+  override def buildScan(requiredColumns: Array[String], filters: Array[Filter]): RDD[Row] = {
+    sparkSession.catalog.listDatabases()
```

Comment:
I was thinking having sparksession.catalog calling these commands, but I guess it is ok this way too ...